designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+# Disruption Controls By Reason
+## TOC and Overview
+1. [Overview](#overview)
+2. [User Scenarios](#user-scenarios)
+3. [Known Requirements](#clarifying-the-requirements-and-behavior)
+   - [Q: Default or Undefined Reason Case Handling](#q-default-or-undefined-reason-case-handling)
+   - [Q: Calculation of Allowed Disruptions in a Multi-Reason World](#q-calculation-of-allowed-disruptions-in-a-multi-reason-world)
+   - [Q: Handling an Unspecified Default Reason](#q-handling-an-unspecified-default-reason)
+   - [Q: Definition of Default in Budgets](#q-definition-of-default-in-budgets)
+4. [API Design](#api-design)
+   - [Approach A: Extending the Budget API to Specify a Reason](#approach-a-extending-the-budget-api-to-specify-a-reason)
+      * [List Approach: Multiple Reasons per Budget](#list-approach-multiple-reasons-per-budget)
+      * [Single Reason Approach: One Reason per Budget](#single-reason-approach-one-reason-per-budget)
+      * [Pros and Cons for Both Approaches](#pros-and-cons-for-both-approaches)
+      * [Preferred Option: List Approach](#preferred-option-list-approach)
+   - [Approach B: Defining Per Reason Controls](#approach-b-defining-per-reason-controls)
+      * [Pros and Cons](#pros-and-cons)
+   - [API Design Conclusion: Preferred Design](#api-design-conclusion-preferred-design)
+
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. Users want to minimize workload disruptions during business hours but still want to be able to delete empty nodes throughout the day.  That is, empty can run all day, while limiting cost savings and upgrades due to drift to non-business hours only.
+
+See Less Made Up Scenarios here: 
+- https://github.com/kubernetes-sigs/karpenter/issues/924 
+- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
+- https://github.com/kubernetes-sigs/karpenter/issues/672
+
+## Clarifying the requirements and behavior 
+**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
+**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 
+
+
+### Q: How should Karpenter handle the default or undefined reason case? 
+If a budget reason is unspecified like budgets[1], we will assume this budget applies to all actions that are not specified 
```

Comment:
Is this the correct behavior? Or should we just assume that it applies to all reasons? Meaning, if this budget was more restrictive, it would just apply to everything and the less restrive budget just wouldn't apply. Think if `budgets[1]` was:

```yaml
budgets:
- nodes: 5
```

Reply:
Based off other convos, I think this is just an artifact from a previous revision

Reply:
I was wondering about this as well. How this is current worded when specifying by reason is an override value which will take precedence over a generic budget even if the generic budget is lower. 

That seems somewhat counter to the "use most restrictive budget", but also does seems potentially alright as I could see in better aligning with real world usage? You might want to have a lower budget generally, but possibly some exemptions to that case for certain reasons? Would be useful to have real world customer desired usage.

Reply:
> I was wondering about this as well. How this is current worded when specifying by reason is an override value which will take precedence over a generic budget even if the generic budget is lower. 

This was my intent. We should use the most restrictive budget per reason. The default budget should only apply to actions not specifed. Is this behavior confusing?  We agreed to not have empty reason mean all actions but instead the default for other actions. 

This budget allows us to say, don't disrupt due to drift or utilization, but feel free to disrupt for emptiness and for expiration as much as you want as I as a user want to be able to have larger disruption for these reasons.

Reply:
2. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes

we still consider all total disruptions, but it would be the min defined by each reason in the budget declaration

Reply:
would you propose the below equation instead? 

AllowedDisruptionByReason = min(minNodeCount[reason], minNodeCount["default"]) - unhealthyNodes - totalDisruptingNodes

Reply:
Let me add this as an equation 3 and we write a thorough pros and cons list going through the 3 equations going through scenarios specified here: https://github.com/kubernetes-sigs/karpenter/pull/936/files#r1503748815

Reply:
After outlining a couple of scenarios, lets go with the inverse of what this says above.  Will push those scenarios 

Take a look at the following budgets. The first budget is easier to reason about and more elegant. Lets ditch this idea. 


# User Story 

Customers who roll out aggressive node image upgrades don't want to roll the nodes due to drift, but want to use node expiration to force users onto compliant node images eventually. The scenario here is an ML Training workload. Ideally we want to let it bake as long as possible, but not at the cost of compromising the security of the cluster.

```yaml 
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenEmpty 
    expireAfter: 7d 
    budgets:
    - nodes: 0
      reasons: [Drifted] 
    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
```

We could also allow a semantic like this
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenEmpty 
    expireAfter: 7d 
    budgets:
    - nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
      reasons: [Expired, Empty]
    - nodes: 0
```
The second semantic may not allow us to define equation 3, but the semantic requires extra work and doesn't follow the original least disruptive principle of the budgets.

Reply:
These are some good detailed comments.

> AllowedDisruptionByReason = min(minNodeCount[reason], minNodeCount["default"]) - unhealthyNodes - totalDisruptingNodes

^ this is a great way of summarizing what I mean.

As the examples you laid out something like the following wouldn't work:
> budgets:
>    \- nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
>      reasons: [Expired, Empty]
>    \- nodes: 0

As by equation 3 all scenarios would be 0. The override scenario pattern, and equation 2 feels the most natural to me, and potentially useful, but wanted to make sure these was at least called out, and discussed as its breaking the "always use the most restrictive" budget if we go with equation 2 over equation 3.

Reply:
A potential way to address this too would be to have a "default" like equation 2, but also support an "All" option? That way if people do want to make sure they have a safe override for example where no disruptions happen they can use the "All" scenario.